### ISSUE TITLE
Pre-Monday Discussion?

### ISSUE THREAD

**Issue by Idi0teque** [obdurodon/dh_course#23]
Hey everyone! Wasn't certain what was meant on the syllabus by "new postings on the course Github site," so I figured I'd make one. (Feel free to delete if this is wrong.)
Anyway, I was wondering about something that came up heavily in my Regex 3 assignment, and that I am 100% certain will also come up with regards to our Magical Realism project. When using Regex find/replace to tag a section from, say, the start of the word SCENE to the start of the next word SCENE, with many characters in between, how would one write a string to do so? I tried things like ^SCENE.+SCENE$, ^SCENE.+?SCENE$, ^SCENE.*$^SCENE, and many other variations, but gave up after 2.5 hours and manually tagged it. Any advice from classmates? (Profs, don't feel obligated to answer!)

**Comment by enb34** [comment]
I got around this by tagging the acts of the play first. Since there are only three scenes in the play we had to markup (one per act), I selected "dot matches all" and found SCENE.+?</act>. This would only work for a play in which there is one act per scene, so this might not be helpful for your other project.

**Comment by djbpitt** [comment]
2.5 hours is too much! When you’ve given it a reasonable try and run out of ideas, this is the right place to post. Here are a few thoughts:

# `^SCENE.+SCENE$`

This matches any line that begins and ends with the string `SCENE` and has at least one character between the two instances of that string. If you turn on "dot matches all", the pattern is no longer constrained to appear in a single line, but it will match greedily, that is, from the very first instance of `SCENE` at the beginning of a line to the very last instance of `SCENE` at the end of a line. 

# ^SCENE.+?SCENE$

This makes it non-greedy, as Emily also suggests above, so that it matches from the first instance of `SCENE` to the next one, rather than to the last one. But this pattern will consume both instances of `SCENE` when it matches them, which means that it will match from the first to the second, but then the second will be gone, so the next match will be from the third to the fourth. Regex does have "look-ahead" and "look-behind" features, which let you define a subpattern without consuming it; you can read about those at https://www.regular-expressions.info/lookaround.html. 

# Milestones

If you want to end one scene and begin another whenever the string `SCENE` appears, you may find it easier to treat it as a milestone. That is, instead of trying to capture and reuse the stuff between instances of `SCENE`, you can replace `SCENE` itself with something like `</scene>\n<scene>`, and then add manually a `<scene>` start tag before the first scene and a `</scene>` end tag after the last one. This milestone approach isn’t as elegant as capturing the stuff you want to wrap because having to fix up the beginning and end manually feels sort of lame. But I often find it easier to think through than look-ahead and look-behind, and I use it where it gets the job done. (And we didn’t expect you to research look-ahead and look-behind for this assignment!)

**Comment by gabikeane** [comment]
This is exactly the right way to make a weekly posting due Monday, you and @enb34 have fulfilled the requirements for this week.  

Did you try checking dot matches all? Generally those larger chunks are easier to do by hand, but we like that you tried to attempt the problem computationally first!

**Comment by brucknerp** [comment]
Another issue with Regex 3: I had trouble with tagging speech, because it would tag more than I needed, sometimes in the middle of stage directions, and I'm wondering how other people went about this search/replace. 
I used \s\s(.+?)(\n\n) because after each character's name there were two spaces, and then a blank link between each character's part, but this also accounted for paragraphs where there were only stage directions as well as some other errors. 

I think I might have done better if I specified the search more to be after a speaker's name, like
</speaker>\s\s(.+?)(\n\n) but I think like posted above, that might require look-ahead functions. 

How did you all handle tagging speech?

**Comment by gmd29** [comment]
I had a similar problem. I think tagging both the speakers and stage directions first might fix it. Then you could search for >(.+?)< which would stop the speech at the stage directions and restart it after the direction.

**Comment by danakaufhold** [comment]
I got around the act/scene weirdness by using the </scene> tag before the </act> tag and the subsequent start of the new act, so when I actually searched for "SCENE," I just added a start-scene tag. (If that makes sense?) It made more sense to me to kill two birds with one stone since there's only one scene per act. This sounds sort of similar to @enb34 's approach.

@brucknerp , I tagged by line spoken rather than speech, so my way of doing this was a little different than yours (because I included the stage directions, even though I tagged them separately). To find each speaker I searched for ^[A-Z][a-z]+\. with dot matches all unchecked, which let me find each line that began with a name. However, this also precluded me from being able to match the names of characters whose names were two words. I didn't mind so much because it wasn't too hard to go back and do separate find/replaces for them, though. I then made each character's name an attribute of the line tag, so my replace was </line>\n<line speaker="character">.

**Comment by JosephDRogers23** [comment]
I completely forgot about the ^ and $ operators to denote the beginning of a line - I ended up creating a generalized pattern, (SECOND ACT)(.*)(THIRD ACT), to tag each act, which I justified only because there were three of them. That would have saved me so much time!

Another issue that I had was tagging each speaker and speech - I ended up creating an attribute named "character" for each of them that had the character, and feel as if I should have associated each speaker with their speech based on some kind of ID. Did anyone address this problem?

**Comment by djbpitt** [comment]
Associating the speeches with the speakers by pointing to an ID in a metadata "personography" is The Right Way to do this sort of thing, primarily because it lets you associate metadata about the character with the speech without having to repeat that same metadata on every speech. You can see an example of this strategy in a professional project (not one of ours) at <http://dh.obdurodon.org/bad-hamlet.xml>. That type of markup would have gone beyond what we expected for this assignment.

Gabi Kirilloff used this approach in her Russian Fairy Tales project to manage gender. She created a metadata personography for each tale, where she declared a unique ID for each character, along with personal information, including gender. When a character spoke in the body of the tale, she used an attribute to point to the relevant ID. She could then *dereference* (= ‘look up’) the gender for each speech by following the pointer from the attribute in the speech to the ID in the personography, and then she could read the gender there. This meant that she could determine which lines were spoken by men and women even though the speeches themselves weren’t tagged for gender; she had to specify the gender of each character only once, in the personography.
